### PR TITLE
Add singularity detection feature

### PR DESCRIPTION
**Pull request by clinssen**
This adds a feature to detect and report on singularity risks in generated propagator matrices.

For a full description of the problem, please see the documentation included in the PR and the following "Summer of Code" project description:

[Singularity Detection and Mitigation in Analytic Integrators.pdf](https://github.com/nest/ode-toolbox/files/9663809/Singularity.Detection.and.Mitigation.in.Analytic.Integrators.pdf)

Most of this work was done by @shraddhajain13.

### CODE REVIEW

**Review by pnbabu — COMMENTED** [review]
Looks good to me overall. Just some minor comments.

File: odetoolbox/singularity_detection.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            mapping from expression that is to be subsituted, to expression to put in its place
+        """
+        for val in sympy.flatten(A):
+            for expr, subs_expr in cond.items():
+                if sympy.simplify(val.subs(expr, subs_expr)) in [sympy.nan, sympy.zoo, sympy.oo]:
+                    return False
+        return True
+
+    @staticmethod
+    def _new_system_matrix(A: sympy.Matrix, cond):  # cond is a tuple here (or a tuple of tuples!)
```
[the remark this answers is not in the document]
Comment (clinssen) [reply odetoolbox/singularity_detection.py]: Cheers, the two unused methods were removed from the class.

**Review by pnbabu — COMMENTED** [review]
Just one more small typo.

**Review by pnbabu — APPROVED** [review]
Thanks!

**Pull request merged**